src/Kernel/Object.class.st

```diff
@@ -1353,6 +1353,14 @@ Object >> isSelfEvaluating [
 	^ self isLiteral
 ]
 
+{ #category : #'self evaluating' }
+Object >> isSelfEvaluating: alreadyVisited [
+
+	(alreadyVisited identityIncludes: self) ifTrue: [ "Recursive structure cannot be a literal" 
+		^ false ].
+	^ self class == Array or: [ self isSelfEvaluating ]
```

Comment:
The type check on Array here seems to make the logic quite complex.
I understand that it is to avoid an infinite reduction on `self isSelfEvaluating` (and that is the core issue of the PR) but it is weird to ask a method in a superclass to have responsibility for the shenanigans of some of its subclasses.

Reply:
No, the check is ok because the "self evaluated" and "literal" printing only works for literals:
```st
#(1 2 3 4 5)
{(1@2). 1}
```

If you evaluate those expressions, Pharo will build Array objects, no subclasses of it.

Reply:
Anyway, we remove the check because is in the entry point method `Array >> isSelfEvaluating`